### DISCUSSION TITLE
How to persist changes to Firebase correctly with QPopupEdit? How to get full record details?

### DISCUSSION THREAD

**Discussion by eugenevk** [quasarframework/quasar#12245] [General - Components / Directives / etc] [answered]
I am using Firebase as my datastore. I have a contacts table and use QPopupEdit for making changes to individual contact fields. When I try to save a change to Firebase (async), I don't get the updated contact record. Therefore, the change doesn't get persisted correctly. I am using the `@save` event to trigger a custom function for saving to Firebase. The `@save` function does get triggered, but only gives me by default the previous and new value of the field I'm changing if I don't use any parameters, but I also need the unique contact identifier.
What is the advised way to do this?

I can't create a Codepen easily for this, because of the credentials.

A screenshot of my contacts table:

![image](https://user-images.githubusercontent.com/19785727/151660815-9300f2bd-a748-49ff-9165-d1748670f2cd.png)

The popup edit:

![image](https://user-images.githubusercontent.com/19785727/151660840-e8c52d46-a46b-4a9d-83eb-2dca6d837c4f.png)

Part of my code for the contacts table:

```html
<q-table
  bordered
  :columns="columns"
  dense
  flat
  :pagination="{ rowsPerPage: 10 }"
  row-key="id"
  :rows="contacts"
  :rows-per-page-options="[10, 25, 50, 100, 0]"
  :table-header-style="{ height: '0px' }"
  :table-header-class="['bg-grey-7', 'text-white']"
>
  <template v-slot:body="props">
    <q-tr :props="props">
      <q-td key="first_name" :props="props" :class="{ 'popup-edit': edit }">
        <span class="flex items-center">
          <span class="q-mr-sm">{{ props.row.first_name }}</span>
          <IconEdit class="hide" />
        </span>
        <q-popup-edit v-model="props.row.first_name" auto-save v-slot="scope" :disable="!edit" @save="saveContactDetails(props.row)">
          <InputText :label="'First name'" v-model="scope.value" :focus="true" @keyup.enter="scope.set" />
        </q-popup-edit>
      </q-td>
...
```

And the function saveContactDetails (here I need to have the contact id or the full contact details with the changes):
```js
const saveContactDetails = async (contact) => {
  // Update the document in Firebase by replacing it entirely
  await setDoc(doc(db, 'contacts', contact.id), contact)
}
```

The columns:

```js
const columns = [
  { name: 'first_name', field: 'first_name', label: 'FIRST NAME', required: true, align: 'left', format: val => `${val}`, sortable: true },
  { name: 'last_name', field: 'last_name', label: 'LAST NAME', required: true, align: 'left', format: val => `${val}`, sortable: true },
  { name: 'position', field: 'position', label: 'POSITION', align: 'left', sortable: true },
  { name: 'gender', field: row => row.gender ? row.gender.label : '', label: 'GENDER', align: 'left', sortable: true },
  { name: 'email', field: 'email', label: 'E-MAIL', align: 'left', sortable: true },
  { name: 'mobile', field: 'mobile', label: 'MOBILE', align: 'left', sortable: true },
  { name: 'has_left', field: 'has_left', label: 'LEFT?', align: 'left', format: val => val ? 'Yes' : 'No', sortable: true },
  { name: 'notes', field: 'notes', label: 'NOTES', align: 'left', sortable: true },
  { name: 'actions', field: null, label: '' }
]
```

**Comment by yusufkandemir** [comment, marked as answer]
First of all, assign a key to your row like this:
```diff
- <q-tr :props="props">
+ <q-tr :key="props.key" :props="props">
```

This would make `props.row` contain the correct details. You probably haven't noticed this yet, because you are trying with the first row. You can test on other records on different pages and see that it contains the wrong row.

As per how to manage updates, you can do something like this:
```js
const updateContactDetails = async (contact, property, newValue) => {
  // Update the document in Firebase by replacing it entirely
  await setDoc(doc(db, 'contacts', contact.id), { ...contact, [property]: newValue })
}
```
then update your `save` event handler
`@save="newValue => updateContactDetails(props.row, 'first_name', newValue)"`

**Reply by eugenevk** [reply]
Awesome @yusufkandemir and thanks for your swift response!

**Comment by eugenevk** [comment]
@yusufkandemir, one additional question though, what does the "newValue => ..." in the save call actually do? Why is that needed instead of just doing the call to updateContactDetails?
